### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ashurakova**

### CODE REVIEW

File: SeaBattle/Program.cs [hunk 1]
```diff
@@ -3,12 +3,29 @@
...
     {
         static void Main(string[] args)
-        {
+        {   
             Console.Title = "Морской бой";
-            
+            FourDeskShip FourDeskShip = new FourDeskShip();
+            ThreeDeskShip ThreeDeskShips = new ThreeDeskShip();
+            TwoDeskShip TwoDeskShips = new TwoDeskShip();
+            OneDeskShip OneDeskShips = new OneDeskShip();
```
Comment (HeorhiSemianeniaiTechArt) [inline SeaBattle/Program.cs]: Where do you use all these ships here?

File: SeaBattleLibrary/Ships/ShipPlacementValidation.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Collections;
+
+namespace SeaBattleLibrary
+{
+    public class ShipPlacementValidation
```
Comment (HeorhiSemianeniaiTechArt) [inline SeaBattleLibrary/Ships/ShipPlacementValidation.cs]: This class is not about ships exactly, so please move it out of Ship folder.

File: SeaBattleLibrary/User/User.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            Console.Write("Ваш выстрел:");
+            string InputtedValue = Console.ReadLine();
+            bool ValidStepMade = false;
+            while (ValidStepMade==false)
+            {
+               
+                int InputtedLetter;
+                int InputtedNumber;
+
+                bool InputValidity = UserStepValidation.StepValidation(InputtedValue, out InputtedLetter, out InputtedNumber);
```
Comment (HeorhiSemianeniaiTechArt) [inline SeaBattleLibrary/User/User.cs]: Please use regex in your validation
image

File: SeaBattleLibrary/User/User.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                        Console.SetCursorPosition(30, 4);
+                        Console.WriteLine();
+
+                    }
+                    ValidStepMade = true;
+                }
+                else
+                {
+                    Console.SetCursorPosition(30, BattleShip.Indent++);
+                    Console.WriteLine("Пожалуйста, введите строчную букву от а до до к и цифру от 1 до 10 в формате 'a1'.");
```
[the remark this answers is not in the document]
Comment (ashurakova) [reply SeaBattleLibrary/User/User.cs]: Fixed

File: SeaBattleLibrary/Ships/ShipPlacementValidation.cs (hunk 2 above)
Comment (ashurakova) [reply SeaBattleLibrary/Ships/ShipPlacementValidation.cs]: Fixed.

File: SeaBattle/Program.cs [hunk 5]
```diff
@@ -3,12 +3,29 @@
...
+                BattleShip.Output(BattleShip.BattleField);
+                if (UserWinValidation.Win() == true)
+                {
+                    GameFinished = true;
+                    break;
+                }
+                user.StepMaking();
+            }
+            Console.SetCursorPosition(30, 1);
+            Console.WriteLine("Спасибо за игру!");
```
[the remark this answers is not in the document]
Comment (ashurakova) [reply SeaBattle/Program.cs]: Fixed

File: SeaBattle/Program.cs (hunk 1 above)
Comment (ashurakova) [reply SeaBattle/Program.cs]: Fixed. ShipPlacement method is used in the Program class for ships creation.

File: SeaBattleLibrary/User/UserStepValidation.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
+﻿using System;
+
+namespace SeaBattleLibrary
+{
+    public class UserStepValidation
+    {
+        public static bool StepValidation(string z, out int ValidLetter, out int ValidNumber)
+        { 
+            char InputtedLetter = z[0];
```
[the remark this answers is not in the document]
Comment (ashurakova) [reply SeaBattleLibrary/User/UserStepValidation.cs]: Fixed

File: SeaBattleLibrary/User/User.cs (hunk 3 above)
Comment (ashurakova) [reply SeaBattleLibrary/User/User.cs]: Fixed. Regex is used for input validation.

File: SeaBattle/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (HeorhiSemianeniaiTechArt) [reply SeaBattle/Program.cs]: good

File: SeaBattleLibrary/User/User.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (HeorhiSemianeniaiTechArt) [reply SeaBattleLibrary/User/User.cs]: good

File: SeaBattle/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (HeorhiSemianeniaiTechArt) [reply SeaBattle/Program.cs]: Can't be verified due to impossibility of game finishing 
image

File: SeaBattle/Program.cs (hunk 1 above)
Comment (HeorhiSemianeniaiTechArt) [reply SeaBattle/Program.cs]: good

File: SeaBattleLibrary/Ships/ShipPlacementValidation.cs (hunk 2 above)
Comment (HeorhiSemianeniaiTechArt) [reply SeaBattleLibrary/Ships/ShipPlacementValidation.cs]: ok

File: SeaBattleLibrary/User/User.cs (hunk 3 above)
Comment (HeorhiSemianeniaiTechArt) [reply SeaBattleLibrary/User/User.cs]: Suitable solution. However, I will close this comment when you fix the comment above about impossibility to finish the game. The problem exists in class User.

File: SeaBattleLibrary/User/UserStepValidation.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
+﻿using System;
+
+namespace SeaBattleLibrary
+{
+    public class UserStepValidation
+    {
+        public static void StepValidation(string z, out int ValidLetter, out int ValidNumber)
+        {
+            int InputtedNumber = Int32.Parse(z.Substring(1)) - 1;
```
Comment (HeorhiSemianeniaiTechArt) [inline SeaBattleLibrary/User/UserStepValidation.cs]: Why do we need to add separate variable for this?

Comment (ashurakova) [reply SeaBattleLibrary/User/UserStepValidation.cs]: Fixed

File: SeaBattle/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashurakova) [reply SeaBattle/Program.cs]: The issue with the impossibility of game finishing is fixed

File: SeaBattleLibrary/User/User.cs (hunk 3 above)
Comment (ashurakova) [reply SeaBattleLibrary/User/User.cs]: The issue with the impossibility of game finishing is fixed

File: SeaBattleLibrary/User/UserStepValidation.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            ValidNumber = InputtedNumber;
+            char InputtedLetter = z[0];
+            int Letter = 0;
```
Comment (HeorhiSemianeniaiTechArt) [inline SeaBattleLibrary/User/UserStepValidation.cs]: Any idea how we could get rid of creating new variable "Letter"?

File: SeaBattleLibrary/User/UserStepValidation.cs (hunk 7 above)
Comment (HeorhiSemianeniaiTechArt) [reply SeaBattleLibrary/User/UserStepValidation.cs]: Fine

File: SeaBattle/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (HeorhiSemianeniaiTechArt) [reply SeaBattle/Program.cs]: ReadKey has been added and it is correct. However, please add some empty lines or somehow show explicitly to user that he wins.
image

File: SeaBattleLibrary/User/User.cs (hunk 3 above)
Comment (HeorhiSemianeniaiTechArt) [reply SeaBattleLibrary/User/User.cs]: damn good!

File: SeaBattle/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashurakova) [reply SeaBattle/Program.cs]: Fixed by clearing console after win

File: SeaBattleLibrary/User/UserStepValidation.cs (hunk 8 above)
Comment (ashurakova) [reply SeaBattleLibrary/User/UserStepValidation.cs]: Fixed

File: SeaBattle/Program.cs [hunk 9]
```diff
@@ -35,13 +35,19 @@ static void Main(string[] args)
...
+                    Console.WriteLine("Спасибо за игру!");
                     GameFinished = true;
                     break;
                 }
                 user.StepMaking();
             }
-            Console.SetCursorPosition(30, 1);
-            Console.WriteLine("Спасибо за игру!");
+            //Console.SetCursorPosition(30, 1);
+            //Console.WriteLine("Спасибо за игру!");
```
Comment (HeorhiSemianeniaiTechArt) [inline SeaBattle/Program.cs]: Please delete commented lines in case they are redundant

Comment (ashurakova) [reply SeaBattle/Program.cs]: Fixed

**Pull request merged**